bpf/lib/conntrack.h

```diff
@@ -873,9 +879,22 @@ ct_lazy_lookup4(const void *map, struct ipv4_ct_tuple *tuple,
 		enum ct_dir dir, enum ct_scope scope, __u32 ct_entry_types,
 		struct ct_state *ct_state, __u32 *monitor)
 {
+	bool is_fragment = false;
+#ifdef ENABLE_IPV4_FRAGMENTS
+	void *data, *data_end;
+	struct iphdr *ip4;
+#endif
+
 	tuple->flags = ct_lookup_select_tuple_type(dir, scope);
 
-	return __ct_lookup4(map, tuple, ctx, l4_off, has_l4_header,
+#ifdef ENABLE_IPV4_FRAGMENTS
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_CT_INVALID_HDR;
+
+	is_fragment = ipv4_is_fragment(ip4);
+#endif
```

Comment:
Could you easily pass a `is_fragment` parameter instead? Revalidating like this will access the wrong header if we're processing the inner payload of an encapsulated packet (ie. what we're doing for XDP-passthrough of GENEVE traffic, where `nodeport_lb4()` takes a `l3_off` != ETH_HLEN).

(uff ok, I had a quick look. For some of the callers that gets quite churny ...)

Reply:
looks like this is becoming a headache :/.

Your idea would be to pass `is_fragment = false` for `ct_lazy_lookup4()` users, and then convert them over 1-by-1 as needed? I think that would be ok, and allow us to make progress on this massive improvement :+1:.